[IV-24-25] Objetivo 5

lmchaves opened the pull request:
* [x] ¿Se han establecido criterios previos de elección de la imagen base, así
como los criterios de búsqueda que han conducido a las imágenes que se están
considerando?
* [x] ¿Se ha tenido en cuenta que la imagen es para testear, y que por tanto
tiene una función específica, tal como testear cómo va nuestra aplicación en la
última versión publicada del lenguage?
* [x] ¿Se ha tenido en cuenta que la versión de la imagen base sea la última?
* [x] ¿Se han tenido en cuenta criterios reales para la elección, y no criterios
irrelevantes como cuanto tiempo se tarda en construirla? (Se va a construir una
vez, se va a usar cientos o miles de veces)
* [x] ¿Se han seguido las buenas prácticas en las órdenes que construyen
   la imagen? ¿Especialmente las relativas a usuarios no privilegiados y demás?
* [x] ¿La imagen incluye sólo y exclusivamente la infraestructura
   necesaria para pasar los tests? Por ejemplo, si se ha elegido la imagen
   "oficial" (lo que hay que hacer sólo en casos muy justificados) ¿se ha
   comprobado que lo que se instale sea sólo y exclusivamente lo necesario para
   nuestro proyecto?
* [x] ¿Se han documentado y enlazado los commits a las imágenes que se
   han testeado?
* [x] ¿Está establecido correctamente como `WORKDIR` el que se va a usar en los
   tests que se van a lanzar?
* [x] ¿Se usa algún directorio adecuado para copiar contenidos del repositorio
en vez del directorio de trabajo?
* [x] ¿Se han establecido claramente los criterios de búsqueda y de aceptación de
   la imagen base?
* [x] ¿Tienes claro cuales son las condiciones en las que tiene que actualizarse
la imagen en DockerHub y has configurado el workflow de acuerdo con ello? En
particular, ¿sabes qué condiciones tienen que darse para que se actualice y qué
condiciones *no* deben darse?
* [x] Si se usan GitHub actions para su publicación, ¿se han considerado
cuidadosamene los criterios (ramas, paths) para que se lancen?
* [x] ¿Se ha comprobado que el ENTRYPOINT del contenedor no escriba nada en el
directorio que monta, ya que no tiene derecho de escritura?
* [x] ¿Se ha abierto un shell en el contenedor una vez construida la imagen y
comprobado que los ficheros que se necesitan están todos incluidos, y que no hay
ficheros que no se necesiten?

lmchaves committed:
feat #41: añado al elección de imagen

lmchaves committed:
feat #42:añado el dockerfile

lmchaves committed:
feat #43: añado el worjflow necesario

lmchaves committed:
Añado el apartado de la imagen

lmchaves committed:
fix #43: no he poddio tenenr el mismo nombre

lmchaves committed:
fix #43: lo pongo bien

hossam1522 commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+RUN apk add --no-cache \
+    make \
+    curl\   
```
¿No viene por defecto wget?

lmchaves replied:
Efectvamente viene por defecto, lo he elminado [b886d07](https://github.com/lmchaves/OrganizarTaller/pull/44/commits/b886d07ece4d042565c81776450f6ea9c24a63cc)

hossam1522 commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+RUN apk add --no-cache \
+    make \
+    curl\   
+    bash 
```
¿Y esto por qué, no viene por defecto?

lmchaves replied:
Es verdad, lo he elminado [b886d07](https://github.com/lmchaves/OrganizarTaller/pull/44/commits/b886d07ece4d042565c81776450f6ea9c24a63cc)

hossam1522 commented on docs/docker_images.md:
```diff
@@ -0,0 +1,33 @@
+
+1. **Imágenes mantenidas:** Preferencia por herramientas activamente mantenidas y actualizadas para evitar problemas de tecnologías obsoletas.
+
+2. Tambien se va a tener en cuenta el peso de las imágenes.
```
El peso base (que es el que estás comparando) realmente es irrelevante. Lo que deberías comparar es el peso final de cada imagen una vez instaladas todas las herramientas necesarias para pasar los tests.

Mario25402 replied:
Me sumo a lo que dice hossam

lmchaves replied:
Lo he cambiado para utiliar el peso final [4305005](https://github.com/lmchaves/OrganizarTaller/pull/44/commits/430500557332d2ae830818d9dab7339752975f6b)

hossam1522 commented on README.md:
```diff
@@ -23,14 +23,21 @@ Para poder realizar los test:
+La documentación sobre la imagen escogida esta en [imagen docker](./docs/docker_images.md)
+
+Para realizar la ejecución del tests:
+docker run -u 1001 -t -v $(pwd):/app/test lmchaveslop/organizartaller   
```
¿Y si quisiera ejecutarlo en local haciendo uso del Dockefile?

lmchaves replied:
Ya lo he añadido [9f70b5d](https://github.com/lmchaves/OrganizarTaller/pull/44/commits/9f70b5df03be3ee9dbad6f2d0dce7a08fa780354)

hossam1522 commented on .github/workflows/docker-build.yml:
```diff
@@ -0,0 +1,29 @@
+    paths:
+      - 'Dockerfile'
+      - 'Makefile'
+      - 'pyproject.toml'
```
Esto no lo estás usando en la imagen, ¿verdad? Entonces, ¿por qué subirías la imagen de nuevo si se actualizan?

lmchaves replied:
Cierto, aqui lo arreglo [288953b](https://github.com/lmchaves/OrganizarTaller/pull/44/commits/288953bff4b40c7910f898ac58a6536739b9bd21)

hossam1522 reviewed (changes requested):
En general tengo que decir que buen trabajo, pero aún tienes que darle una vueltecilla. ¡Ánimo!

Mario25402 commented on docs/docker_images.md:
```diff
@@ -0,0 +1,33 @@
+
+## Conlusión
+
+Hay varias opciones para elegir una imagen base para el proyecto. Las imágenes como Ubuntu, Debian o Alpine solo incluyen el sistema operativo, y todas están activamente mantenidas. Alpine destaca por ser la más ligera, pero presenta el inconveniente de que debemos instalar Python y sus dependencias adicionales, lo que puede aumentar el tiempo de construcción y el tamaño final.
```
Realmente que el tiempo de construcción sea más o menos no es tan relevante (dentro de unos límites claro) dado a que la imagen se contruirá hasta que consigas la versión final, a partir de ahí, solo vas a ejecutarla, la instalación de python tampoco será muy pesada. Opino que puedes quitar lo del tiempo de construcción, lo del tamaño final sí me parece adecuado.

lmchaves replied:
Perfecto, lo he quitado y he mantenido el peso final: [864375c](https://github.com/lmchaves/OrganizarTaller/pull/44/commits/864375c01e8a8df758fddee8b8d2af1bdc4a1149)

Mario25402 reviewed (commented):
Hola, te dejo dos puntualizaciones, buen trabajo!!

JLombar commented on DOCKER_USER:
```diff
@@ -0,0 +1 @@
+lmchaveslop
```
No estoy realmente seguro, he buscado pero no entiendo qué es este archivo. Esto no debería estar en un gitignore? No lo usas en nada no?

hossam1522 replied:
Ese archivo es por

> Si por cualquier razón no se ha podido usar en Docker Hub el mismo nick que en GitHub, se creará un fichero DOCKER_USER en el directorio principal con el mismo.

http://jj.github.io/IV/documentos/proyecto/5.Docker#entrega-del-material-evaluable-en-este-objetivo

lmchaves replied:
Exacto, lo he añadido al ignore [5ea39f3](https://github.com/lmchaves/OrganizarTaller/pull/44/commits/5ea39f3b441a89cb488de4d0dd39f05cdeedcb72)

JJ replied:
Agh, pero ¿para qué lo añades al ignore? Si ya lo he marcado yo como resuelto. ¿De qué sirve ponerlo en el .gitignore *si ya está en el repositorio*?

lmchaves replied:
Tienes razo, ya lo he solucionado

JLombar commented on docs/docker_images.md:
```diff
@@ -0,0 +1,33 @@
+
+## Criterios de evaluación
+
+Para seleccionar el gestor de dependencias más adecuado para este proyecto, se han considerado los siguientes criterios:
```
Gestor de dependencias?

lmchaves replied:
Ya esta solucionado [864375c](https://github.com/lmchaves/OrganizarTaller/pull/44/commits/864375c01e8a8df758fddee8b8d2af1bdc4a1149)

JJ replied:
Si está solucionado porque hayáis cambiado, que pongáis en qué comit lo habéis cambiado no añade tanto trabajo, pero indica que no entendéis por qué se debe poner el commit. Por favor, no lo hagáis.

JLombar reviewed (commented):
Poco más a parte de lo que te comentan los compañeros

lmchaves committed:
fix #42: elimino bash y curl

Ya que wget esta por defecto y bash tambien

lmchaves committed:
fix #41: añado la ejecución en local

lmchaves committed:
fix #41:elimno pyproject.toml

Ya que no es usada en la imagen

lmchaves committed:
fix #41:utilizo peso final en lugar de peso base

lmchaves committed:
fix #41: elmino de la conclusión el tiempo de construccion

lmchaves committed:
fix #43: añado al gitignore

lmchaves commented:
Gracias a todos por los comentraios, ya los he solucionado, si véis que todavía hay errores comentarlo

lmchaves committed:
fix #43: no es necesario ya que ya esta en el repo

hossam1522 commented on .github/workflows/docker-build.yml:
```diff
@@ -0,0 +1,28 @@
+  push:
+    paths:
+      - 'Dockerfile'
+      - 'Makefile'
```
No lo estás usando dentro de la imagen, ¿verdad?

lmchaves replied:
Es verdad, ya lo he cambiado

JJ replied:
Tienes que resolver *todos* los comentarios.

hossam1522 reviewed (commented):
Un pequeño apunte.

lmchaves committed:
fix #43 elimo makefile, ya que se usa dentro de la imagen

hossam1522 reviewed (approved):
Poco más que añadir por mi parte.

JJ commented on .github/workflows/docker-build.yml:
```diff
@@ -0,0 +1,27 @@
+  push:
+    paths:
+      - 'Dockerfile'
+
```
¿No hay ningún fichero que incluyas en el Dockerfile que necesite también monitorizarse para actualizar la imagen?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,26 @@
+FROM python:alpine
```
No nos hemos calentado mucho la cabeza, ¿eh?

lmchaves replied:
Se que un comentario, se resuelve con cambio en el código, pero no se muy bien porque python:alpine no vale, yo la elegí porque de las opciones que probé es la menos pesada de final (con todfas las herramientas)

JJ replied:
¿Has mirado qué hay instalado en esa versión y cómo lo instala?
Recuerda que mirar *sólo* las opciones con Python ya instalado no es la única opción.

JJ reviewed (changes requested):
El Dockerfile es prácticamente el mismo que el de @juanmaaf salvo por el cambio de la imagen base. Realmente me sentiría más cómodo si hubiera alguna variación. Nunca hay una sola forma de hacer las cosas, y la forma que se parezca más a la ya hecha nunca va a ser la adecuada.

lmchaves committed:
fix #42:cambio la manera de tratar uv

JLombar reviewed (approved):
Perdona, no vi el review request la semana pasada y me acabo de dar cuenta. Si cambias lo que te propone JJ por mi parte está aprobado

lmchaves commented:
@JLombar he cambiado el dockerfile, y me podiras echar una mano con el cambio del docker-build.yml, ya que actualmente en el Dockerfile no veo ningun archivo sea necesario monitorizar actualmente, ya qeu tanto Makefile,pyproject.tom y DOCKER_USER hacen puente con los archivos locales, y no se muy bien si sería neceasrio añadir alguno mas

juanmaaf reviewed (commented):
Viendo que has resuelto los comentarios de las distintas revisiones, tienes mi apto. Sin embargo, si miras la pestaña Files Changed verás que hay comentarios sin resolver, por lo que así no se puede revisar ni aprobar el objetivo. Un comentario se resuelve con un commit en la línea comentada.

JLombar commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+WORKDIR /app/test
+
+CMD ["make", "test"]
```
Es con entrypoint no cmd

lmchaves replied:
Ya lo he cambiado

JLombar commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+
+USER userTest
+
+RUN python3 -m pip install --user pipx && \
```
Esto de usar pip no sé como de bien está realmente. No sabes que versión instalas ni nada, lo veo un poco ambiguo. Creo que deberías descargar a instakar uv con wget o algo así antes que esto. 

Si usas pip se descarga la versión más compatible en tu sistema, si el mio por lo que sea tiene otra tenemos 2 versiones diferentes del paquete y no creo que sea la forma de hacer esto

lmchaves replied:
Ya lo he modificado

JLombar reviewed (commented):
Perdona por no haberte respondido antes.

JLombar commented:
@lmchaves si crees que está bien explica a JJ el por qué respondiendo el comentario no? Puede ser que estés en lo correcto. Sin embargo si haces un cambio en el makefile no se debe ver reflejado en la imagen de dockerhub?

lmchaves committed:
fix #42 elimino el uso de pip

lmchaves committed:
fix #42 uso entrypoint

lmchaves committed:
fix #43 añado makefile

lmchaves commented:
Ya lo he arreglado @JLombar

lmchaves committed:
closes #43 elmino ruido

lmchaves committed:
closes #43 elmino ruido

lmchaves committed:
fix #42: elimno el uso de pip

lmchaves committed:
fix #41 y #42 cambio la imagen

Cambio python:alp
ine ya que no poedemos controalar que versión de python viene instalado, por ejemplo en la uitlma versión python:alpine3
.21 podemos ver en el Dockerfile que utiliza python 3.11.11, en mi caso sería válido ya que en el pyproject.toml requiro
 de superior a 3.10, pero si no fuese el caso podria dar a error

lmchaves committed:
closes #41: modifico la conclusion

lmchaves committed:
closes #42: no es necesario

lmchaves commented:
@JJ ya esta listo para revision

JJ reviewed (changes requested):
Como te avisó @juanmaaf, tienes que resolver todos los comentarios.

lmchaves committed:
closes #43 no se utiliza en la imagen

lmchaves commented:
ya esta resuelto @JJ

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,26 @@
+
+RUN apk add --no-cache \
+    make \
+    python3 
```
¿Qué versión de Python estás añadiendo aquí?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,26 @@
+    make \
+    python3 
+
+WORKDIR /app/test
```
¿Dónde es necesario establecer esto? ¿Los comandos siguientes lo usan?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,26 @@
+
+RUN mkdir -p /home/userTest/.cache && \
+    chmod -R a+w /home/userTest/.cache && \
+    mkdir -p /home/userTest/.venv && \
```
¿Por qué es necesario esto?

lmchaves replied:
Ya lo he explicado correcatmente en #47

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,26 @@
+    mkdir -p /home/userTest/.venv && \
+    chmod -R a+w /home/userTest/.venv 
+
+ENV PATH="/home/userTest/.local/bin:$PATH"
```
¿Qué hay en ese path?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,26 @@
+
+RUN wget -qO- https://astral.sh/uv/install.sh | sh
+
+ENV UV_CACHE_DIR=/home/userTest/.cache/uv
```
Si estas líneas apuntan al #42 que es un issue genérico y no presenta ningún problema y se añaden con un commit que dice "Añado Dockerfile" debo asumir que las has copiado o que son innecesarias porque no estás diciendo qué problema resuelven. Además, sin saber qué problema resuelven no se sabe si hace falta una, las dos, o ninguna de ellas.
Te hago notar, por otro lado, que estás usando repetidamente `/home/userTest`. Si tuvieras que cambiar eso tendrías que cambiarlos todos.

lmchaves replied:
Ya he credo el issue correcto en #46

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,26 @@
+
+RUN adduser -D -h /home/userTest userTest
+
+USER userTest
```
¿Este nombre de usuario sigue la mejor práctica en cuanto al uso de mayúsculas y minúsculas? ¿Incluir `user`, por otro lado, crees que aporta alguna información?

JJ commented on docs/docker_images.md:
```diff
@@ -0,0 +1,35 @@
+2. Tambien se va a tener en cuenta el peso final de las imágenes(con todas las herramientas necesarias).
+
+
+## [Ubuntu](https://hub.docker.com/_/ubuntu)
```
¿Estás comparando imágenes sin python con imágenes con Python en cuestión de peso?

lmchaves replied:
Aqui tengo una duda, el problema no es que la opción no tenga pyhton no? si no el criterio con el que lo comparo?

JJ replied:
No, el problema es exactamente lo que te estoy diciendo. Estás metiendo en el mismo saco imágenes con Python que imágenes con él. No creo que sean comparables a nivel de peso, por ejemplo, al menos que realmente instales Python en las imágenes sin él y compares.

lmchaves replied:
El peso final del archivo del que hablo (el de la iamgen), incluye python a las que no lo teinen, perdón por no espeficiarlo claramente, ya lo he resuelto

JJ commented on docs/docker_images.md:
```diff
@@ -0,0 +1,35 @@
+La última actualización fue de hace unos días, siendo una herramieta activa y mantenida, con un peso mucho menor al de las anteriores, de tan solo 5MB de imagen base, sin funcionalides extra siendo mínima y eficiente. Pero con las herramientas,tiene un peso final de **114MB**.
+
+## [Python](https://hub.docker.com/_/python)
+Ya que estoy utilizando python, tenemos que ver esta opción en la que ya nos viene instalado python, pero hay diversas opciones, de las ya comentadas anteriormente como Alpine(muy ligera) peso final **85MB**, debian:bookworm-slim con peso final **163MB**, python:bullseye-slim con un peso final de **161MB**. Con la última acutalzición de días, por lo que esta mantenida.
```
Resuelve typos por favor.

JJ commented on docs/docker_images.md:
```diff
@@ -0,0 +1,35 @@
+
+Hay varias opciones para elegir una imagen base para el proyecto. Las imágenes como Ubuntu, Debian o Alpine solo incluyen el sistema operativo, y todas están activamente mantenidas. Alpine destaca por ser la más ligera, pudiendo instalar Python y sus dependencias adicionales.
+
+Entre las opciones de python, la mejor es la que incluye alpine ya que es la menos pesada, pero tien el inconveniente de que no sabemos que versión de python estamos instalando, y si es la que necesitamos para nuestro proyecto
```
¿Cómo es que no lo sabes?

JJ reviewed (commented):
Como he dicho en clase, lo más importante es aprender a crear la imagen especificando los problemas que se presenten y demostrando por tanto que se está haciendo paso a paso y siguiendo la metodología adecuada.
También conviene que evites errores a la hora de teclear.

lmchaves committed:
close #42: realizo correctamente el problema

lmchaves committed:
feat #46: agrego la solucion de cache

lmchaves committed:
fix #46 solución al problema de cache

lmchaves committed:
feat #47: arreglo el problem de venv

Lo que tenemos que hacer es que cree el venv en un lugar donde tenga los permisos

lmchaves committed:
fix #42: pongo un correcto nombre

lmchaves committed:
fix #42: esablezco una versión de python

lmchaves committed:
closes #43: lo establezco correctamente

lmchaves committed:
closes #42: corrigo la version de python

lo pongo correctamente mediante ENV

lmchaves committed:
fix #41: corrigo typos

lmchaves committed:
fix #41: modifico la conclusion

lmchaves committed:
closes #41: especifico el peso

lmchaves commented:
@JJ ya esta listo para revisión

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,29 @@
+
+ENV PATH="/home/pruebatest/.local/bin:$PATH"
+
+ENV PYTHON_VERSION 3.11.11
```
De veras, Luis, te trato de corregir las cosas con rapidez para que puedas sacar adelante esto, pero ya te dije antes que por qué estabas intalando python3 de los repositorios y has eliminado el comentario. No has creado ningún issue relativo a la necesidad de instalación de Python y como hacerlo, pero lo que es peor, estás usando aquí una versión que, primero, no sabemos por qué la necesitas (porque el issue que la pone es genérico, el #42) y segundo, tampoco sabemos por qué es esta y no otra. Pero es que aunque lo supiéramos *siempre es un error poner versiones específicas en la imagen de Docker".

JJ reviewed (changes requested):
Hay también un comentario que no has resuelto. Te agradecería que lo hicieras, porque realmente deberías al menos organizar las imágenes base de otra forma.
Por otro lado, el problema fundamental sigue siendo el mismo: debéis incluir el la imagen sólo lo que resuelva problemas específicos, exactamente igual que en los otros objetivos. Que en este caso no salgan directamente de una HU no hace que sean diferentes.
Hace horas te dije en #42 que debías hacer que cada issue fuera un problema. Añadiste nuevos commits, cuando la forma correcta de hacerlo es borrar el código  con un commit que cierre el issue como innecesario o inadecuado, y crear nuevos issues **que representen problemas**

Hasta ahora has trabajado correctamente y has acabado superando los objetivos. Por favor, sigue haciéndolo.

lmchaves committed:
closes #42:este issue es inadecuado

lmchaves committed:
closes #42: no hay que usar versiones específcas

lmchaves committed:
fix #41: dejo claro las imagen base

lmchaves committed:
fix #41: arreglo la conclusión

lmchaves committed:
feat #48: añadmios las dependecias

con esta imagen nos garantizamos la ultima versión de python, instalamos make, configuramos el path para que uv se instale correctmaente

lmchaves committed:
closes #46:soluciono el problema

para que no nos de este error tenenmos que garantizar que actua en una carpeta con los correctos permisos

lmchaves committed:
closes #47: soluciono el error

ya que uv intenta crear un venv en app/test, donde no tiene permisos, tenemos que inidcarle en que lugar crear los venv

lmchaves committed:
fix #48 pongo correctamente el nombre

lmchaves committed:
closes #41: eligo la imagen definitiva

lmchaves committed:
closes #48:hago los cambios para debian

lmchaves commented:
Lo primero gracias por la paciencia, juraría que actualmente he creado los issues para que representen un problema, y solucionado con el código, he dejado de utilizar versiones especificas en el dockerfile, y en princiio resuelto todos los comentarios @JJ

JJ commented on docs/docker_images.md:
```diff
@@ -0,0 +1,39 @@
+
+
+
+## Conlusión
```
Mira lo que pone.

JJ commented on docs/docker_images.md:
```diff
@@ -0,0 +1,39 @@
+## Imágenes con python incluido 
+
+### [Python](https://hub.docker.com/_/python)
+Ya que estoy utilizando python, tenemos que ver esta opción en la que ya nos viene instalado python, pero hay diversas opciones, de las ya comentadas anteriormente como Alpine(muy ligera) peso final **85MB**, debian:bookworm-slim con peso final **163MB**, python:bullseye-slim con un peso final de **161MB**. Con la última actualización  de días, por lo que esta mantenida.
```
Pone "esta"

JJ reviewed (approved):
:+1: 
Voy a optar por aprobar por los cambios que has hecho, pero realmente deberís haber considerado más tipos de imágenes con Python. Has puesto una sola. Sigues teniendo muchos errores al teclear. Por favor corrige los que puedas.

lmchaves committed:
closes #41: corrijo los typo
